src/common/dav/tx.c

```diff
@@ -439,27 +468,35 @@ dav_tx_begin(dav_obj_t *pop, jmp_buf env, ...)
 		}
 
 		/* inherits this value from the parent transaction */
-		struct tx_data *txd = PMDK_SLIST_FIRST(&tx->tx_entries);
+		struct tx_data *txd = DAV_SLIST_FIRST(&tx->tx_entries);
 
 		failure_behavior = txd->failure_behavior;
 
 		VALGRIND_START_TX;
 	} else if (tx->stage == DAV_TX_STAGE_NONE) {
+		struct umem_store *store = &pop->do_store;
+
+		DAV_DBG("");
+
 		VALGRIND_START_TX;
 
 		dav_hold_clogs(pop);
 		operation_start(pop->undo);
 
 		VEC_INIT(&tx->actions);
-		PMDK_SLIST_INIT(&tx->tx_entries);
+		DAV_SLIST_INIT(&tx->tx_entries);
 
 		tx->ranges = ravl_new_sized(tx_range_def_cmp,
 			sizeof(struct tx_range_def));
-
-		tx->pop = pop;
-
 		tx->first_snapshot = 1;
 
+		err = store->stor_ops->so_wal_reserv(store, &pop->do_utx->utx_id);
```

Comment:
Please be aware that wal_reserve() could yield, so another ULT could see the tx->stage is STAGE_NONE and try to start TX by the per-thread tx too.

The wal_reserve() should be called at the very beginning of this function. (before any tx->stage checks).

Reply:
Actually wal_reserve() should not break apart the ULT switch variable check and its setting which are tx->stage check and tx->stage = DAV_TX_STAGE_WORK; in this case. Then the UTL context switch at wal_reserve() would not break the control flow.